pkg/apis/application/v1alpha1/app_project_types.go

```diff
@@ -336,10 +336,15 @@ func (proj *AppProject) RemoveFinalizer() {
 	setFinalizer(&proj.ObjectMeta, ResourcesFinalizerName, false)
 }
 
-func globMatch(pattern string, val string, separators ...rune) bool {
+func globMatch(pattern string, val string, allowNegation bool, separators ...rune) bool {
+	if allowNegation && strings.HasPrefix(pattern, "!") {
+		return !glob.Match(pattern[1:], val, separators...)
+	}
+
 	if pattern == "*" {
```

Comment:
Should we handle the weird case where someone does `!*`? Maybe need some unit tests of just this function.

Reply:
I was thinking on how to best do that. Do we consider that `!*` is deny-all and reject any other rules? Or do we consider that `!*` is in itself an invalid rule and reject rules based upon that?

Reply:
Hm. I think we should probably consider it invalid. I can't imagine a case where someone actually needs to block all destinations for a project.

Reply:
Alright, I added a check to `ValidateProject` to enforce that the rule cannnot be created in the first place. Let me know if that works for you.

Reply:
Also added a final note in the docs about `!*` being an invalid rule, let me know if that works so I can tick off the documentation from the checklist :)